middleware.js

```diff
@@ -1,19 +1,38 @@
 import { NextResponse } from "next/server";
+import { verifyAuth, isAdmin } from "@/lib/auth";
 
 
 // This function can be marked `async` if using `await` inside
-export function middleware(request) {
-
+export async function middleware (request) {
   // there are some public paths and there are some protected paths
   // the public path should not be visible when the user has the token
   // the private path should not be visible when the user doesn't have the token
-
   const path = request.nextUrl.pathname;
   const isPublicPath = path === "/login" || path === "/signup";
-  console.log("middlware called for path=", path)
 
   const token = request.cookies.get("token")?.value || ""; // check if the token exists
 
+   // validate the user is authenticated
+   const verifiedToken = await verifyAuth(request);
+  
+    //validate the user is authorized
+   const isAdmin_ = isAdmin(verifiedToken);
+  console.log("isAdmin_ = ", isAdmin_)
+
+  //https://nextjs.org/docs/messages/returning-response-body-in-middleware
+  //https://medium.com/sopra-steria-norge/how-to-write-actual-api-middleware-for-next-js-2a38355f6674
+  if(path.startsWith("/admin") && isAdmin_) {
+    console.log("admin is allowed for this route")
+    return NextResponse.next()
+  } else if(path.startsWith("/admin") && !isAdmin_) {
+    console.log("ACCESS DENIED: only admin is allowed for this route")
+    //let response = NextResponse.json({error: "ACCESS DENIED: only admin is allowed for this route"}, {status: 301}) //301 -> redirect
+    //console.log("DENIED response1 = ", response)
+    //return response;
+    return NextResponse.redirect(new URL("/", request.nextUrl));
```

Comment:
Non admin user was able to access /admin route, because `NextResponse.redirect` was never returned.

Fix: added `return` before `NextResponse.redirect`